### PR TITLE
feat(communities): User is able to cancel membership request

### PR DESCRIPTION
**Pull request by borismelnik**
Part of: #7072

[Status-go PR](https://github.com/status-im/status-go/pull/2915)

### What does the PR do

Add methods to cancel membership requests to community

### Affected areas

Communities

### StatusQ checklist

- [ ] add documentation if necessary (new component, new feature)
- [ ] update sandbox app
  - in case of new component, add new component page
  - in case of new features, add variation to existing component page
  - nice to have: add it to the demo application as well
- [ ] test changes in both light and dark theme?

### Screenshot of functionality (including design for comparison)

- [x] I've checked the design and this PR matches it

https://user-images.githubusercontent.com/82511785/197172029-0cb4f83b-37c5-4df8-8d03-d4cb08dd0409.mov

UPDATE:
@caybro @PascalPrecht I update ui logic as expect:

https://user-images.githubusercontent.com/82511785/197408822-42345c19-f216-46a5-8501-543a6d7caa66.mov

### CODE REVIEW

**Comment by caybro** [comment]
Oh, so you cancel the request by clicking on the "Pending..." button :o ? That's quite unexpected, wdyt @John-44

**Comment by borismelnik** [comment]
> Oh, so you cancel the request by clicking on the "Pending..." button :o ? That's quite unexpected, wdyt @John-44

Yes, unfortunately, I can't find a mockup where it should be (or I just blind :) )

**Comment by PascalPrecht** [comment]
I don't think we have a design for cancelling such requests. But I agree with @caybro that we'll need a call to action button for cancellation, while "Pending" is more of a label of the current state

**Comment by borismelnik** [comment]
> I don't think we have a design for cancelling such requests. But I agree with @caybro that we'll need a call to action button for cancellation, while "Pending" is more of a label of the current state

I agreed. It was just for demonstration and a reason for this discussion :)

File: src/app_service/service/community/service.nim [hunk 1]
```diff
@@ -194,11 +198,19 @@ QtObject:
...
           var community = self.joinedCommunities[membershipRequest.communityId]
-          community.pendingRequestsToJoin.add(membershipRequest)
-          self.joinedCommunities[membershipRequest.communityId] = community
-          self.events.emit(SIGNAL_COMMUNITY_EDITED, CommunityArgs(community: community))
+          if (membershipRequest.state == RequestToJoinType.Pending.int):
+            community.pendingRequestsToJoin.add(membershipRequest)
+            self.joinedCommunities[membershipRequest.communityId] = community
+            self.events.emit(SIGNAL_COMMUNITY_EDITED, CommunityArgs(community: community))
+
+            self.events.emit(SIGNAL_NEW_REQUEST_TO_JOIN_COMMUNITY, CommunityRequestArgs(communityRequest: membershipRequest))
```
Comment (PascalPrecht) [inline src/app_service/service/community/service.nim:206]: Just to be sure, this line belongs into the if-case here?

File: src/app_service/service/community/service.nim [hunk 2]
```diff
@@ -1197,6 +1234,19 @@ QtObject:
 
     self.joinedCommunities[communityId] = community
 
+  proc cancelRequestToJoinCommunity*(self: Service, communityId: string) =
+    try:
+      var i = 0
+      for communityRequest in self.myCommunityRequests:
+        if (communityRequest.communityId == communityId):
+          discard status_go.cancelRequestToJoinCommunity(communityRequest.id)
```
Comment (PascalPrecht) [inline src/app_service/service/community/service.nim:1242]: What if this errors out?

**Review by PascalPrecht — APPROVED** [review]
Overall LGTM

Just the comment about the pending state label becoming a call to action button.
And some comments to clarify

File: src/app_service/service/community/service.nim (hunk 1 above)
Comment (borismelnik) [reply src/app_service/service/community/service.nim:205]: fixed :)

File: src/app_service/service/community/service.nim (hunk 2 above)
Comment (borismelnik) [reply src/app_service/service/community/service.nim]: Fixed :)

**Review by MishkaRogachev — APPROVED** [review]
Looks good!
If it would be merged after https://github.com/status-im/status-go/pull/2886, cancelling the membership request should also be supported in the AC notifications

File: src/app_service/service/community/service.nim [hunk 3]
```diff
@@ -194,11 +198,19 @@ QtObject:
...
-          community.pendingRequestsToJoin.add(membershipRequest)
-          self.joinedCommunities[membershipRequest.communityId] = community
-          self.events.emit(SIGNAL_COMMUNITY_EDITED, CommunityArgs(community: community))
+          if (membershipRequest.state == RequestToJoinType.Pending.int):
+            community.pendingRequestsToJoin.add(membershipRequest)
+            self.joinedCommunities[membershipRequest.communityId] = community
+            self.events.emit(SIGNAL_COMMUNITY_EDITED, CommunityArgs(community: community))
+            self.events.emit(SIGNAL_NEW_REQUEST_TO_JOIN_COMMUNITY, CommunityRequestArgs(communityRequest: membershipRequest))
+
+          if (membershipRequest.state == RequestToJoinType.Canceled.int):
```
Comment (endulab) [inline src/app_service/service/community/service.nim:207]: elif

File: src/app_service/service/community/service.nim [hunk 4]
```diff
@@ -1203,6 +1240,24 @@ QtObject:
...
+          let response = status_go.cancelRequestToJoinCommunity(communityRequest.id)
+          if (not response.error.isNil):
+            let msg = response.error.message & " communityId=" & communityId
+            error "error while cancel membership request ", msg
+            return
+          self.myCommunityRequests.delete(i)
+
+        i.inc()
+      
+      self.events.emit(SIGNAL_REQUEST_TO_JOIN_COMMUNITY_CANCELED, Args())
```
Comment (endulab) [inline src/app_service/service/community/service.nim:1257]: @borismelnik this signal is for the future ? It's not handled anywhere.

**Review by endulab — APPROVED** [review]

File: src/app_service/service/community/service.nim [hunk 5]
```diff
@@ -1203,6 +1240,24 @@ QtObject:
 
     self.joinedCommunities[communityId] = community
 
+  proc cancelRequestToJoinCommunity*(self: Service, communityId: string) =
+    try:
+      var i = 0
+      for communityRequest in self.myCommunityRequests:
+        if (communityRequest.communityId == communityId):
+          let response = status_go.cancelRequestToJoinCommunity(communityRequest.id)
```
Comment (MishkaRogachev) [inline src/app_service/service/community/service.nim:1248]: Here should be emited AC notification: `self.activityCenterService.parseACNotificationResponse(response)`

File: src/app_service/service/community/service.nim (hunk 4 above)
Comment (borismelnik) [reply src/app_service/service/community/service.nim:1257]: yes

**Pull request merged**